<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Cnidaria by sauloal</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Cnidaria</h1>
        <ul>
          <li><a href="https://github.com/sauloal/cnidaria/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/sauloal/cnidaria/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/sauloal/cnidaria">View On <strong>GitHub</strong></a></li>
        </ul>
        <p></p>

        <p class="view"><a href="https://github.com/sauloal/cnidaria">View the Project on GitHub <small>sauloal/cnidaria</small></a><img src="https://travis-ci.org/sauloal/cnidaria.svg?branch=master"/></p>
      </header>
      <section>
        <h1><p>CNIDARIA: fast, reference-free phylogenomic clustering</p></h1>

<p>Motivation: Identification of biological specimens is a major requirement 
for a range of applications. Reference-free methods analyse unprocessed 
sequencing data without relying on prior knowledge, but these do not scale 
to arbitrarily large genomes and arbitrarily large phylogenetic distances.</p>

<p>Results: We present Cnidaria, a practical tool for clustering genomic and 
transcriptomic data with no limitation on ge-nome size or phylogenetic 
distances. We successfully simultaneously clustered 169 genomic and 
transcriptomic datasets from 4 kingdoms, achieving 100% accuracy at 
supra-species level and 78% accuracy for species level.</p>

<p>Availability and Implementation: Cnidaria is written in C++ and Python and 
is available at <a href="http://www.ab.wur.nl/cnidaria">http://www.ab.wur.nl/cnidaria</a>.</p>

<p>Contact: <a href="mailto:sauloal@gmail.com">sauloal@gmail.com</a></p>

<p>Supplementary information: Supplementary data are available at 
Bioinformatics online.</p>


<h1>How to run</h1>
<h2>Docker</h2>
<p>At src/docker you can find the docker recipes.</p>

<h2>Install</h2>
<h3>To install the system requirements:</h3>
<pre>
$ sudo bash INSTALL.1.requirements
</pre>

<h3>Minimum requirements:</h3>
<p><strong>Ubuntu >= 13.4</strong></p>
<pre>
lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.2 LTS
Release:        14.04
Codename:       trusty
</pre>

<p><strong>libboost >= 1.50</strong></p>
<pre>
$ dpkg -s libboost-dev | grep 'Version'
Version: 1.54.0.1ubuntu1

$ cat /usr/include/boost/version.hpp | grep "BOOST_LIB_VERSION"
//  BOOST_LIB_VERSION must be defined to be the same as BOOST_VERSION
#define BOOST_LIB_VERSION "1_54"
</pre>

<p><strong>gcc >= 4.8</strong></p>
<pre>
$gcc -v  
gcc version 4.8.2 (Ubuntu 4.8.2-19ubuntu1)
</pre>



<h3>To re-compile code:</h3>
<pre>
$ bash INSTALL.2.compile
</pre>

<p>The compilation creates a python library called _cnidariapy.so, a shared library for python.</p>


<h2>Setup</h2>
<h3>how to generate jf database files:</h3>
<p>In analysis/data/converters you can find helper scripts.</p>
<ul>
  <li>fasta2jf.sh converts FASTA fo JF. </li>
  <li>cram2jf.sh converts CRAM fo JF (requires recent samtools).</li>
  <li>bam2jf.sh converts BAM fo JF (requires samtools). </li>
</ul>

<h3>Go to the folder where your fasta files are</h3>
<p>Symlink jopts, fasta2jf.sh, cram2jf.sh, bam2jf.sh and run.sh from analysis/data/converters.</p>
<p>Edit jopts to match your kmer requirements and the jellyfish/samtools executables
analysis/data/converters/jopts</p>
<p>The jopts file should have the path to your jellyfish and samtools executables and the
kmer size. Samtools are only needed if you are importing BAM files</p>
<pre>
JELLYFISH=/home/user/cnidaria/src/libs/Jellyfish/bin/jellyfish
SAMTOOLS=/home/user/bin/samtools/samtools-1.1/samtools

KSIZE=15

if [[ "$KSIZE" == 31 ]]; then
MER_SIZE=31
HASH_SIZE=4G
fi

if [[ "$KSIZE" == 21 ]]; then
MER_SIZE=21
HASH_SIZE=1G
fi

if [[ "$KSIZE" == 17 ]]; then
MER_SIZE=17
HASH_SIZE=512M
fi

if [[ "$KSIZE" == 15 ]]; then
MER_SIZE=15
HASH_SIZE=256M
fi

if [[ "$KSIZE" == 11 ]]; then
MER_SIZE=11
HASH_SIZE=128M
fi



COUNTER_LEN=7
OUT_COUNTER_LEN=1
JCMD="${JELLYFISH} count -m ${MER_SIZE} -s ${HASH_SIZE} -t 5 -F 300 --disk --counter-len=${COUNTER_LEN} --out-counter-len=${OUT_COUNTER_LEN} --canonical"
ulimit -Sn 4096
</pre>

<p>Run ./run.sh</p>

<p>It will search for all fasta/cram/bam under the current folder and convert them to JF.</p>

<h3>Create a filelist.csv file</h3>
<p>Create a filelist.csv file, a tab delimited file containg the name of your JF files and their "pretty" name:</p>
<pre>
/home/user/cnidaria/data/input/spp1.fas.21.jf	Species 01
/home/user/cnidaria/data/input/spp2.fas.21.jf	Species 02
/home/user/cnidaria/data/input/spp3.fas.21.jf	Species 03
/home/user/cnidaria/data/input/spp4.fas.21.jf	Species 04
/home/user/cnidaria/data/input/spp5.fas.21.jf	Species 05
</pre>

<h3>Create a test_def.csv file</h3>
<p>Create a test_def.csv file, a tab delimited file containing the run names</p>
<pre>
test01	/home/user/cnidaria/data/input/spp1.fas.21.jf
test01	/home/user/cnidaria/data/input/spp2.fas.21.jf
test01  /home/user/cnidaria/data/input/spp3.fas.21.jf
test01  /home/user/cnidaria/data/input/spp4.fas.21.jf
test01  /home/user/cnidaria/data/input/spp5.fas.21.jf
</pre>

<h2>Run</h2>
<h3>Automatically</h3>
<h4>Create a Makefile</h4>
<p>Create a Makefile for your runs by using scripts/gen_mkfile.py.</p>
<p>It will create a Makefile for your analysis calling all programs in the
correct order for: split, calculate, merge, generate statistics and plot 
graphs in one go</p>
<pre>
scripts/gen_mkfile.py -h
#usage: gen_mkfile.py [-h] [-thr [NUM_THREADS]] [-min [MINVAL]]
#                     [-se [SAVE_EVERY]] [-me] [-nm] [-ec] [-nem]
#                     file_list def_file out_dir kmer_size num_pieces

# ./gen_mkfile.py file_list def_file out_folder kmer_size number_of_pieces
./gen_mkfile.py /home/user/cnidaria/data/filelist.csv \
  /home/user/cnidaria/data/test_def.csv \
  /home/user/cnidaria/data/output 21 20
</pre>



<h4>Run Makefile</h4>
<p>Go to the output folder ( /home/user/cnidaria/data/output ) and run:</p>
<pre>
  Make all
</pre>

<p>Or do a specific run:</p>
<pre>
  Make test01
</pre>

<p>the makefile contains:</p>
<pre>
# Run Cnidaria for each of the pieces
ulimit -c unlimited && time /home/user/cnidaria/scripts/cnidaria.py --export-complete \
--num-pieces 20 --piece-num 1 --outfile test01 /home/user/cnidaria/data/input/spp{1,2,3,4,5}.fas.21.jf

# Merge all the pieces
ulimit -c unlimited && time /home/user/cnidaria/scripts/cnidaria.py --export-complete \
--num-pieces 20 --merge-only --outfile test01 /home/user/cnidaria/data/input/spp{1,2,3,4,5}.fas.21.jf
  
# Create statistics of the results
/home/user/cnidaria/scripts/cnidaria_stats.py test01_0001_0020.json /home/user/cnidaria/data/filelist.csv
...
/home/user/cnidaria/scripts/cnidaria_stats.py test01_0020_0020.json /home/user/cnidaria/data/filelist.csv
  
# Verify if the merging occured well
/home/user/cnidaria/scripts/verify_csvs.py test01_00{01,02,...,19,20}_0020.json.csv test01.json.csv
  
# Merge statistics in single files
cat test*.json.csv               > test01.all.csv
cat test01_*_0020.json.count.csv > test01.all.count.csv

# Convert trees into PNG
/home/user/cnidaria/scripts/newick_to_png.py test01_00{01,02,...,19,20}_0020.json.no_scale.jaccard_dissimilarity.nj
</pre>

<p>The output files are:</p>
<pre>
# Results of individual pieces
test01_0001_0020.cne
test01_0001_0020.cnm
test01_0001_0020.json
test01_0001_0020.json.count.csv
test01_0001_0020.json.csv
test01_0001_0020.json.no_scale.jaccard_dissimilarity.matrix
test01_0001_0020.json.no_scale.jaccard_dissimilarity.nj
test01_0001_0020.json.no_scale.jaccard_dissimilarity.nj.png
test01_0001_0020.json.no_scale.jaccard_dissimilarity.nj.tree

# Global results
test01.all.count.csv
test01.all.csv
test01.cne
test01.cnm
test01.json
test01.json.count.csv
test01.json.csv
test01.json.no_scale.jaccard_dissimilarity.matrix
test01.json.no_scale.jaccard_dissimilarity.nj
test01.json.no_scale.jaccard_dissimilarity.nj.png
test01.json.no_scale.jaccard_dissimilarity.nj.tree
</pre>

<p><strong>CNE</strong> - a Cnidaria Complete Database file containing both k-mer and presence/absence list</p>
<p><strong>JSON</strong> - a Cnidaria Summary Database file containing only the matrix of shared/total k-mers counts</p>
<p><strong>MATRIX</strong> - a jaccard distance matrix calculated from the JSON file</p>
<p><strong>NJ</strong> - a NEWICK neighbour-joining tree based in MATRIX</p>
<p><strong>PNG</strong> - a tree image beased in NJ</p>
<p><strong>TREE</strong> - an ASCII art tree based in NJ</p>

<h4>To create HTML report</h4>
<pre>
/home/user/cnidaria/scripts/report/stats_report.py test01.json \
/home/user/cnidaria/data/filelist.csv \
test01.json.no_scale.jaccard_dissimilarity.nj

#which will generate:
#test01_kmer_stats_report.html
</pre>


<h3>Manually</h3>
<p>Alternatively, you can run scripts/cnidaria.py, the main program, for the analysis.</p>
<pre>
scripts/cnidaria.py -h
#usage: cnidaria.py [-h] [-out [OUT_FILE]] [-thr [NUM_THREADS]] [-min [MINVAL]]
#                   [-se [SAVE_EVERY]] [-np [NUM_PIECES]] [-pn [PIECE_NUM]]
#                   [-n] [-d] [-me] [-nm] [-ec] [-nem]
#                   infiles [infiles ...]
</pre>





      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/sauloal">sauloal</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-5291039-10");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
